In [57]:
# Importing the required libraries.

import pandas as pd
import numpy as np


In [58]:
# Loading the dataset into pandas data frame

data = pd.read_csv("London Housing Data.csv")
data.head()

date            area  average_price       code  houses_sold  \
0  1/1/1995  city of london          91449  E09000001         17.0   
1  2/1/1995  city of london          82203  E09000001          7.0   
2  3/1/1995  city of london          79121  E09000001         14.0   
3  4/1/1995  city of london          77101  E09000001          7.0   
4  5/1/1995  city of london          84409  E09000001         10.0   

   no_of_crimes  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

## Initial Data set analysis

In [59]:
data.shape

(13549, 6)

In [60]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13549 entries, 0 to 13548
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           13549 non-null  object 
 1   area           13549 non-null  object 
 2   average_price  13549 non-null  int64  
 3   code           13549 non-null  object 
 4   houses_sold    13455 non-null  float64
 5   no_of_crimes   7439 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 635.2+ KB


In [61]:
data.describe()

average_price    houses_sold  no_of_crimes
count   1.354900e+04   13455.000000   7439.000000
mean    2.635197e+05    3893.994129   2158.352063
std     1.876175e+05   12114.402476    902.087742
min     4.072200e+04       2.000000      0.000000
25%     1.323800e+05     247.000000   1623.000000
50%     2.229190e+05     371.000000   2132.000000
75%     3.368430e+05    3146.000000   2582.000000
max     1.463378e+06  132163.000000   7461.000000

In [62]:
data.isnull().sum()

date                0
area                0
average_price       0
code                0
houses_sold        94
no_of_crimes     6110
dtype: int64

In [63]:
data.isnull().mean()

date             0.000000
area             0.000000
average_price    0.000000
code             0.000000
houses_sold      0.006938
no_of_crimes     0.450956
dtype: float64

## Convert the datatype of "date" column to date-time_format.

In [64]:
data.head()

date            area  average_price       code  houses_sold  \
0  1/1/1995  city of london          91449  E09000001         17.0   
1  2/1/1995  city of london          82203  E09000001          7.0   
2  3/1/1995  city of london          79121  E09000001         14.0   
3  4/1/1995  city of london          77101  E09000001          7.0   
4  5/1/1995  city of london          84409  E09000001         10.0   

   no_of_crimes  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [65]:
data["date"].dtype

dtype('O')

In [66]:
data["date"] = pd.to_datetime(data["date"])

In [67]:
data["date"].dtype

dtype('<M8[ns]')

In [68]:
data.dtypes

date             datetime64[ns]
area                     object
average_price             int64
code                     object
houses_sold             float64
no_of_crimes            float64
dtype: object

#### Answer - Job done

## Add a column year in the dataframe, which contains years only

In [69]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [70]:
data["year"] = data["date"].dt.year

In [71]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

#### Answer - Job done

## Add a new column "month" as 2nd column in the data frame, which contains month only.

In [72]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

In [73]:
data.insert(loc = 1 , column = "month" , value = data["date"].dt.month)

In [74]:
data.head()

date  month            area  average_price       code  houses_sold  \
0 1995-01-01      1  city of london          91449  E09000001         17.0   
1 1995-02-01      2  city of london          82203  E09000001          7.0   
2 1995-03-01      3  city of london          79121  E09000001         14.0   
3 1995-04-01      4  city of london          77101  E09000001          7.0   
4 1995-05-01      5  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

#### Answer - Job Done

## Remove the columns "year" and "month" from the dataframe.

In [75]:
data.drop(columns = ["year" , "month"] , axis = 1 , inplace = True)

In [76]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

#### Answer - Job done

## Show all the records where "no_of_crimes" is 0. And how many such records are there ?

In [77]:
pd.set_option("display.max_rows" , 500)

In [78]:
data["no_of_crimes"].value_counts()

0.0       104
1956.0     11
2039.0     11
2098.0     10
2276.0     10
         ... 
898.0       1
3874.0      1
2010.0      1
3545.0      1
5450.0      1
Name: no_of_crimes, Length: 2669, dtype: int64

In [79]:
data[data["no_of_crimes"]==0]

date            area  average_price       code  houses_sold  \
72  2001-01-01  city of london         284262  E09000001         24.0   
73  2001-02-01  city of london         198137  E09000001         37.0   
74  2001-03-01  city of london         189033  E09000001         44.0   
75  2001-04-01  city of london         205494  E09000001         38.0   
76  2001-05-01  city of london         223459  E09000001         30.0   
77  2001-06-01  city of london         242435  E09000001         36.0   
78  2001-07-01  city of london         230700  E09000001         40.0   
79  2001-08-01  city of london         231632  E09000001         30.0   
80  2001-09-01  city of london         229806  E09000001         30.0   
81  2001-10-01  city of london         263887  E09000001         26.0   
82  2001-11-01  city of london         346458  E09000001         15.0   
83  2001-12-01  city of london         362204  E09000001         17.0   
84  2002-01-01  city of london         344239  E09000001         30.0   
85  2002-02-01  city of london         284015  E09000001         12.0   
86  2002-03-01  city of london         253697  E09000001         47.0   
87  2002-04-01  city of london         245313  E09000001         31.0   
88  2002-05-01  city of london         227281  E09000001         30.0   
89  2002-06-01  city of london         253130  E09000001         25.0   
90  2002-07-01  city of london         262428  E09000001         26.0   
91  2002-08-01  city of london         274016  E09000001         51.0   
92  2002-09-01  city of london         277047  E09000001         49.0   
93  2002-10-01  city of london         271941  E09000001         19.0   
94  2002-11-01  city of london         280283  E09000001         28.0   
95  2002-12-01  city of london         262197  E09000001         44.0   
96  2003-01-01  city of london         261645  E09000001         32.0   
97  2003-02-01  city of london         263992  E09000001         20.0   
98  2003-03-01  city of london         291969  E09000001         31.0   
99  2003-04-01  city of london         295174  E09000001         30.0   
100 2003-05-01  city of london         284860  E09000001         38.0   
101 2003-06-01  city of london         267208  E09000001         84.0   
102 2003-07-01  city of london         260204  E09000001         48.0   
103 2003-08-01  city of london         245449  E09000001         46.0   
104 2003-09-01  city of london         243184  E09000001         38.0   
105 2003-10-01  city of london         252980  E09000001         33.0   
106 2003-11-01  city of london         277515  E09000001         42.0   
107 2003-12-01  city of london         305556  E09000001         38.0   
108 2004-01-01  city of london         326913  E09000001         98.0   
109 2004-02-01  city of london         301092  E09000001         16.0   
110 2004-03-01  city of london         272731  E09000001         20.0   
111 2004-04-01  city of london         269514  E09000001         25.0   
112 2004-05-01  city of london         280062  E09000001         22.0   
113 2004-06-01  city of london         297421  E09000001         38.0   
114 2004-07-01  city of london         298293  E09000001         34.0   
115 2004-08-01  city of london         307094  E09000001         46.0   
116 2004-09-01  city of london         305126  E09000001         27.0   
117 2004-10-01  city of london         289912  E09000001         24.0   
118 2004-11-01  city of london         333075  E09000001         13.0   
119 2004-12-01  city of london         331161  E09000001         14.0   
120 2005-01-01  city of london         330363  E09000001         17.0   
121 2005-02-01  city of london         288867  E09000001         18.0   
122 2005-03-01  city of london         304851  E09000001         18.0   
123 2005-04-01  city of london         317398  E09000001         32.0   
124 2005-05-01  city of london         311442  E09000001         20.0   
125 2005-06-01  city of london         282866  E09000001         33.

In [80]:
len(data[data["no_of_crimes"]==0])

104

#### Answer - There are total 104 records where number of crimes is 0.

## What is the maximum and minimum "average_price" per year in England?

In [81]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [82]:
# Adding the year column which was dropped in the last question.
data["year"] = data["date"].dt.year

In [83]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

In [84]:
# Filtering the data where area is england.
d = data[data["area"]=="england"]
d.head()

date     area  average_price       code  houses_sold  \
13248 1995-01-01  england          53203  E92000001      47639.0   
13249 1995-02-01  england          53096  E92000001      47880.0   
13250 1995-03-01  england          53201  E92000001      67025.0   
13251 1995-04-01  england          53591  E92000001      56925.0   
13252 1995-05-01  england          53678  E92000001      64192.0   

       no_of_crimes  year  
13248           NaN  1995  
13249           NaN  1995  
13250           NaN  1995  
13251           NaN  1995  
13252           NaN  1995

In [85]:
d.groupby(by = "year")["average_price"].max()

year
1995     53901
1996     55755
1997     61564
1998     65743
1999     75071
2000     84191
2001     95992
2002    119982
2003    138985
2004    160330
2005    167244
2006    182031
2007    194764
2008    191750
2009    174136
2010    180807
2011    177335
2012    180129
2013    188544
2014    203639
2015    219582
2016    231922
2017    242628
2018    248620
2019    250410
2020    247355
Name: average_price, dtype: int64

In [86]:
d.groupby(by = "year")["average_price"].min()

year
1995     52788
1996     52333
1997     55789
1998     61659
1999     65522
2000     75219
2001     84245
2002     96215
2003    121610
2004    139719
2005    158572
2006    166544
2007    181824
2008    165795
2009    159340
2010    174458
2011    173046
2012    174161
2013    176816
2014    188265
2015    202856
2016    220361
2017    231593
2018    240428
2019    243281
2020    247355
Name: average_price, dtype: int64

## What is the maximum and minimum no. of crimes recorded per area?

In [87]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

In [88]:
data.groupby(by = "area")["no_of_crimes"].max()

area
barking and dagenham      2049.0
barnet                    2893.0
bexley                    1914.0
brent                     2937.0
bromley                   2637.0
camden                    4558.0
city of london              10.0
croydon                   3263.0
ealing                    3401.0
east midlands                NaN
east of england              NaN
enfield                   2798.0
england                      NaN
greenwich                 2853.0
hackney                   3466.0
hammersmith and fulham    2645.0
haringey                  3199.0
harrow                    1763.0
havering                  1956.0
hillingdon                2819.0
hounslow                  2817.0
inner london                 NaN
islington                 3384.0
kensington and chelsea    2778.0
kingston upon thames      1379.0
lambeth                   4701.0
lewisham                  2813.0
london                       NaN
merton                    1623.0
newham                    3668.0
north

In [89]:
data.groupby(by = "area")["no_of_crimes"].min()

area
barking and dagenham      1217.0
barnet                    1703.0
bexley                     860.0
brent                     1850.0
bromley                   1441.0
camden                    2079.0
city of london               0.0
croydon                   2031.0
ealing                    1871.0
east midlands                NaN
east of england              NaN
enfield                   1635.0
england                      NaN
greenwich                 1513.0
hackney                   1870.0
hammersmith and fulham    1323.0
haringey                  1536.0
harrow                     937.0
havering                  1130.0
hillingdon                1445.0
hounslow                  1529.0
inner london                 NaN
islington                 1871.0
kensington and chelsea    1347.0
kingston upon thames       692.0
lambeth                   2381.0
lewisham                  1675.0
london                       NaN
merton                     819.0
newham                    2130.0
north

## Show the total count of records of each area, where average price is less than 100000.

In [90]:
data.head()

date            area  average_price       code  houses_sold  \
0 1995-01-01  city of london          91449  E09000001         17.0   
1 1995-02-01  city of london          82203  E09000001          7.0   
2 1995-03-01  city of london          79121  E09000001         14.0   
3 1995-04-01  city of london          77101  E09000001          7.0   
4 1995-05-01  city of london          84409  E09000001         10.0   

   no_of_crimes  year  
0           NaN  1995  
1           NaN  1995  
2           NaN  1995  
3           NaN  1995  
4           NaN  1995

In [91]:
# Filtering the data where average price is less than 100000.
data[data["average_price"]<100000]["area"].value_counts()

north east              112
north west              111
yorks and the humber    110
east midlands            96
west midlands            94
england                  87
barking and dagenham     85
south west               78
east of england          76
newham                   72
waltham forest           64
bexley                   64
lewisham                 62
havering                 60
south east               59
greenwich                59
croydon                  57
sutton                   54
enfield                  54
hackney                  53
redbridge                52
southwark                48
tower hamlets            47
outer london             46
hillingdon               44
hounslow                 41
lambeth                  41
brent                    40
london                   39
merton                   35
haringey                 33
bromley                  33
inner london             31
ealing                   31
kingston upon thames     30
harrow              

## Thank You

# Sumit Bhatia